File: backend/app/schemas/ai_reading_assistant.py
"""
AI 阅读助手数据模型

FUTURE-AI-READING-ASSISTANT-1 P2 实现
定义阅读计划草案和相关的数据结构
"""

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


# ==================== 枚举 ====================

class ReadingGoalType(str, Enum):
    """阅读目标类型"""
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    YEARLY = "yearly"


class SuggestionType(str, Enum):
    """建议类型"""
    CONTINUE = "continue"   # 继续阅读
    START = "start"         # 开始新书
    FINISH = "finish"       # 完成当前
    PAUSE = "pause"         # 暂停建议


class SuggestionPriority(str, Enum):
    """建议优先级"""
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


# ==================== 阅读目标 ====================

class ReadingGoal(BaseModel):
    """阅读目标"""
    goal_type: ReadingGoalType = Field(..., description="目标类型")
    target_count: int = Field(1, description="目标数量")
    current_count: int = Field(0, description="当前进度")
    media_types: list[str] = Field(default_factory=list, description="适用媒体类型")
    deadline: Optional[str] = Field(None, description="截止日期")
    description: str = Field("", description="目标描述")


# ==================== 阅读建议 ====================

class ReadingSuggestion(BaseModel):
    """阅读建议"""
    suggestion_type: SuggestionType = Field(..., description="建议类型")
    media_type: str = Field(..., description="媒体类型: novel / manga / audiobook")
    item_id: Optional[int] = Field(None, description="书籍/漫画 ID")
    title: str = Field(..., description="标题")
    author: Optional[str] = Field(None, description="作者")
    reason: str = Field("", description="推荐理由")
    priority: SuggestionPriority = Field(SuggestionPriority.MEDIUM, description="优先级")
    estimated_time: Optional[str] = Field(None, description="预估阅读时间")
    current_progress: Optional[str] = Field(None, description="当前进度（如有）")


# ==================== 阅读计划草案 ====================

class ReadingPlanDraft(BaseModel):
    """阅读计划草案"""
    summary: str = Field("", description="总体说明")
    goals: list[ReadingGoal] = Field(default_factory=list, description="阅读目标")
    suggestions: list[ReadingSuggestion] = Field(default_factory=list, description="阅读建议")
    stats_context: dict[str, Any] = Field(default_factory=dict, description="统计背景")
    insights: list[str] = Field(default_factory=list, description="阅读洞察")
    generated_at: Optional[str] = Field(None, description="生成时间")
    
    @classmethod
    def fallback_draft(cls, error_message: str) -> "ReadingPlanDraft":
        """生成回退草案（解析失败时使用）"""
        return cls(
            summary=f"AI 输出解析失败: {error_message[:100]}",
            insights=["阅读计划生成失败，请稍后重试或检查 AI 服务状态。"],
            generated_at=datetime.now().isoformat(),
        )


# ==================== 请求/响应 ====================

class ReadingPlanRequest(BaseModel):
    """阅读计划请求"""
    prompt: Optional[str] = Field(
        None,
        description="用户自然语言描述（可选）"
    )
    focus: Optional[str] = Field(
        None,
        description="聚焦类型: all / novel / manga / audiobook"
    )
    goal_type: Optional[str] = Field(
        None,
        description="目标类型: daily / weekly / monthly"
    )


class ReadingPlanResponse(BaseModel):
    """阅读计划响应"""
    success: bool = Field(True, description="是否成功")
    plan: Optional[ReadingPlanDraft] = Field(None, description="阅读计划草案")
    error: Optional[str] = Field(None, description="错误信息")


class PresetReadingPrompt(BaseModel):
    """预设阅读提示词"""
    id: str
    title: str
    prompt: str
    description: str
    focus: Optional[str] = None


class PresetReadingPromptsResponse(BaseModel):
    """预设阅读提示词响应"""
    prompts: list[PresetReadingPrompt] = Field(default_factory=list)


# ==================== 辅助函数 ====================

def priority_to_score(priority: SuggestionPriority) -> int:
    """优先级转分数（用于排序）"""
    scores = {
        SuggestionPriority.HIGH: 0,
        SuggestionPriority.MEDIUM: 1,
        SuggestionPriority.LOW: 2,
    }
    return scores.get(priority, 1)

File: frontend/src/api/aiReadingAssistant.ts
/**
 * AI 阅读助手 API
 *
 * FUTURE-AI-READING-ASSISTANT-1 P4 实现
 */

import axios from 'axios'

const API_BASE = '/api/ai/reading-assistant'

// ==================== 类型定义 ====================

export type ReadingGoalType = 'daily' | 'weekly' | 'monthly' | 'yearly'
export type SuggestionType = 'continue' | 'start' | 'finish' | 'pause'
export type SuggestionPriority = 'high' | 'medium' | 'low'

export interface ReadingGoal {
  goal_type: ReadingGoalType
  target_count: number
  current_count: number
  media_types: string[]
  deadline?: string
  description: string
}

export interface ReadingSuggestion {
  suggestion_type: SuggestionType
  media_type: string
  item_id?: number
  title: string
  author?: string
  reason: string
  priority: SuggestionPriority
  estimated_time?: string
  current_progress?: string
}

export interface ReadingPlanDraft {
  summary: string
  goals: ReadingGoal[]
  suggestions: ReadingSuggestion[]
  stats_context: Record<string, any>
  insights: string[]
  generated_at?: string
}

export interface ReadingPlanRequest {
  prompt?: string
  focus?: string // 'all' | 'novel' | 'manga' | 'audiobook'
  goal_type?: string // 'daily' | 'weekly' | 'monthly'
}

export interface ReadingPlanResponse {
  success: boolean
  plan?: ReadingPlanDraft
  error?: string
}

export interface PresetReadingPrompt {
  id: string
  title: string
  prompt: string
  description: string
  focus?: string
}

export interface PresetReadingPromptsResponse {
  prompts: PresetReadingPrompt[]
}

export interface ServiceStatus {
  enabled: boolean
  service: string
  version: string
  features: string[]
  supported_media_types: string[]
}

// ==================== API 方法 ====================

export const aiReadingAssistantApi = {
  /**
   * 生成阅读计划草案
   */
  async generate(request: ReadingPlanRequest): Promise<ReadingPlanResponse> {
    const response = await axios.post<ReadingPlanResponse>(`${API_BASE}/generate`, request)
    return response.data
  },

  /**
   * 获取预设提示词
   */
  async getPresetPrompts(): Promise<PresetReadingPromptsResponse> {
    const response = await axios.get<PresetReadingPromptsResponse>(`${API_BASE}/preset-prompts`)
    return response.data
  },

  /**
   * 检查服务状态
   */
  async getServiceStatus(): Promise<ServiceStatus> {
    const response = await axios.get<ServiceStatus>(`${API_BASE}/status`)
    return response.data
  },
}

export default aiReadingAssistantApi
